utt/signature-processor/src/sigProcessor.cpp

```diff
@@ -0,0 +1,181 @@
+                                   const auto entries = jobs_.getEntries();
+                                   for (const auto& [id, entry] : entries) {
+                                     (void)id;
+                                     if (entry.start_time == 0) continue;
+                                     auto now = std::chrono::steady_clock::now().time_since_epoch().count();
+                                     if (now - entry.start_time >= entry.job_timeout_ms_) {
+                                       onJobTimeout(entry);
+                                     }
+                                   }
+                                 });
+}
+void SigProcessor::processRCMSignature(uint64_t sigId,
+                                       const libutt::api::types::CurvePoint& pidHash,
+                                       const libutt::api::Commitment& rcm1,
+                                       const libutt::api::types::CurvePoint& s2,
+                                       std::shared_ptr<libutt::api::Registrator> signer,
+                                       uint64_t job_timeout) {
+  jobs_.registerEntry(sigId, [signer, rcm1, s2, pidHash](uint64_t rid, const libutt::api::types::Signature& sig) {
+    (void)sig;
+    (void)rid;
+    return true;
+  });
+  auto partial_sig = signer->signRCM(pidHash, s2, rcm1);
+  std::vector<uint64_t> sigIds = {sigId};
+  handleSigJob(sigIds, {partial_sig.second}, job_timeout);
+}
+
+void SigProcessor::handleSigJob(std::vector<uint64_t> sigIds,
+                                const std::vector<libutt::api::types::Signature>& partial_sigs,
+                                uint64_t job_timeout) {
+  if (sigIds.size() != partial_sigs.size()) return;
+  for (size_t i = 0; i < sigIds.size(); i++) {
+    auto sigId = sigIds.at(i);
+    auto& partial_sig = partial_sigs.at(i);
+    auto& jobEntry = jobs_.get(sigId);
+    auto& validator = jobs_.getJobValidator(sigId);
+    std::vector<uint16_t> invalids;
+    for (auto& [id, psig] : jobEntry.partial_sigs) {
+      if (!validator((uint16_t)id, psig)) invalids.push_back((uint16_t)id);
+    }
+    for (auto id : invalids) jobEntry.partial_sigs.erase((uint32_t)id);
+
+    jobEntry.partial_sigs.emplace((uint32_t)repId_, partial_sig);
+    jobEntry.job_id = sigId;
+    jobEntry.start_time = std::chrono::steady_clock::now().time_since_epoch().count();
+    jobEntry.job_timeout_ms_ = job_timeout;
+    jobs_.put(sigId, jobEntry);
+    auto job_coordinator = getJobCoordinator(sigId);
+    uint16_t rep = repId_;
+    messages::PartialSigMsg msg(rep, partial_sig, sigId);
+    msgsCommunicator_->sendAsyncMessage((uint64_t)job_coordinator, msg.body(), msg.size());
+  }
+}
+SigProcessor::~SigProcessor() { timers_.cancel(timeout_handler_); }
+void SigProcessor::onReceivingNewPartialSig(uint64_t sigId,
+                                            uint16_t sig_source,
+                                            const libutt::api::types::Signature& partial_sig) {
```

Comment:
Worth adding assert for sigID%cluster_size == rep_id

Reply:
replicas that are not the leader also may get partial signatures from others (in case they hit the timeout)

Reply:
cool ,thanks